### ISSUE TITLE
storage layout customization

### ISSUE THREAD

**Issue by sakulstra** [foundry-rs/foundry#2138]
### Component

Forge

### Describe the feature you would like

I would be nice if `forge inspect <file> storage-layout --pretty` would had an option like to set a width or column specific constraints. The change is probably trivial as https://docs.rs/comfy-table/6.0.0/comfy_table/struct.Column.html which is used to generate the table supports this out of the box, but i'm not sure what would be the best interface to accustom user needs without introducing to much configuration complexity.

The first thing i thought off might be to just allow a `--sizes=80,100,120` parameter which can set individual fixed sizes on columns. Wdyt?

### Additional context

I'm currently using foundry to generate a diff between two proxy implementations.
To generate a nice human readable diff using `storage-layout --pretty` comes in quite handy. One issue i'm facing though is that due to addition or renaming of variables the table layout might change causing git diff algorithm to assume everything has changed.

```diff
-|----------------------------------+----------------------------------------------------------------------------------------+------+--------+-------|
-| _decimals                        | uint8                                                                                  | 5    | 0      | 1     |
+|-------------------------------------+--------------------------------------------------------------+------+--------+-------|
+| _decimals                           | uint8                                                        | 5    | 0      | 1     |
```

**Comment by onbjerg** [comment]
I'd recommend for the purposes of diffing that you don't use the `--pretty` flag. We can add a `--sizes` option but it's only going to apply to that singular field so I lean more towards not adding it simply to try and keep the CLI simple when we can

**Comment by onbjerg** [comment]
>Could you elaborate? As i understand the comfy_table docs it should be possible to provide min/max/absolute size for each column.

The `forge inspect` command supports displaying almost any field in the artifacts generated by Forge. Some of them respond to `--pretty`, but not all of them are put in a table when `--pretty` is set. I can't really recall any others that are put into a table other than `storage_layout` so this flag would really only be for that single field

>I'm doing the actual diff without the --pretty, but when generating the diff to be embedded inside a md the --pretty option would be nicer.

I'm not sure I understand, - you diff without `--pretty` but the diff in the issue is with `--pretty`? Can you elaborate a bit? :smile:

**Comment by sakulstra** [comment]
Sure that comment might have been a bit confusing :sweat_smile: 

Currently i use `forge inspect <file> storage-layout` because
a) --pretty [was not including](https://github.com/foundry-rs/foundry/pull/2087) the filenames
b) with `--pretty` in some circumstances (max type/label string length changed) the output is not properly dffable (the issue described here)

I'd rather use `--pretty` (if b was resolved) as it's easier to embed inside a longer markdown file as the tabular layout being easier to read and less loc then the json type output (which contains the `ast` property and the `types` in the array in root).

**Comment by sakulstra** [comment]
@onbjerg not sure if i should better create a new issue for this, but as it's somewhat related:
Would you be open to:
a) alter the default --pretty print or
b) add a `--markdown` flag

To make the output table compatible with markdown?
In essence this would mean to change from:

```
+----------------------------------+----------------------------------------------------------------------------------------+------+--------+-------+----------+
| Name                             | Type                                                                                   | Slot | Offset | Bytes | Contract |
+==============================================================================================================================================================+
| _balances                        | mapping(address => uint256)                                                            | 0    | 0      | 32    | Contract |
|----------------------------------+----------------------------------------------------------------------------------------+------+--------+-------+----------|
| _allowances                      | mapping(address => mapping(address => uint256))                                        | 1    | 0      | 32    | Contract |
```

To sth like:
```
| Name                             | Type                                                                                   | Slot | Offset | Bytes | Contract |
| -------------------------------- | -------------------------------------------------------------------------------------- | ---: | -----: | ----: | -------- |
| _balances                        | mapping(address => uint256)                                                            | 0    | 0      | 32    | Contract |
| _allowances                      | mapping(address => mapping(address => uint256))                                        | 1    | 0      | 32    | Contract |
```

Which will render as:
| Name                             | Type                                                                                   | Slot | Offset | Bytes | Contract |
| -------------------------------- | -------------------------------------------------------------------------------------- | ---: | -----: | ----: | -------- |
| _balances                        | mapping(address => uint256)                                                            | 0    | 0      | 32    | Contract |
| _allowances                      | mapping(address => mapping(address => uint256))                                        | 1    | 0      | 32    | Contract |

**Comment by KholdStare** [comment]
Hi @onbjerg, any thoughts on this issue? I think there are a few overlapping concerns raised here, but one that we have run into is reliable storage diffs. Without `--pretty` comparing the json output usually results in `astId` mismatches, even though the storage is identical. With `--pretty` we may get some false differences due to formatting. Maybe something formatting agnostic like `--csv` would be simplest for the reliable diff use-case.

**Comment by GopherJ** [comment]
seems `--pretty` doesn't have struct fields shown in the table right?

![image](https://user-images.githubusercontent.com/33961674/216248065-e97b36ee-85b3-466e-842a-90cbcbecf58a.png)

**Comment by sakulstra** [comment]
closing as this was resolved in #3666
